src/Compilers/Shared/BuildServerConnection.cs

```diff
@@ -75,107 +29,156 @@ internal sealed class BuildServerConnection
+                // server simply isn't running.
+                return true;
             }
 
-            if (buildPaths.TempDirectory == null)
+            try
             {
-                throw new ArgumentException(nameof(buildPaths));
+                var request = BuildRequest.CreateShutdown();
+
+                // Don't create the server when sending a shutdown request. That would defeat the 
+                // purpose a bit.
+                var response = await RunServerBuildRequestAsync(
+                    request,
+                    pipeName,
+                    timeoutOverride,
+                    tryCreateServerFunc: (_, _) => false,
+                    logger,
+                    cancellationToken).ConfigureAwait(false);
+
+                if (response is ShutdownBuildResponse shutdownBuildResponse)
+                {
+                    if (waitForProcess)
+                    {
+                        try
+                        {
+                            var process = Process.GetProcessById(shutdownBuildResponse.ServerProcessId);
+                            process.WaitForExit();
+                        }
+                        catch (Exception)
+                        {
+                            // There is an inherent race here with the server process.  If it has already shutdown
+                            // by the time we try to access it then the operation has succeed.
+                        }
+                    }
+
+                    return true;
+                }
+
+                return wasServerRunning(pipeName) == false;
+            }
+            catch (Exception)
+            {
+                // If the server was in the process of shutting down when we connected then it's reasonable
+                // for an exception to happen.  If the mutex has shutdown at this point then the server 
+                // is shut down.
+                return wasServerRunning(pipeName) == false;
             }
 
+            // Was a server running with the specified session key during the execution of this call?
+            static bool? wasServerRunning(string pipeName)
+            {
+                string mutexName = GetServerMutexName(pipeName);
+                return WasServerMutexOpen(mutexName);
+            }
+        }
+
+        internal static Task<BuildResponse> RunServerBuildRequestAsync(
+            BuildRequest buildRequest,
+            string pipeName,
+            string clientDirectory,
+            ICompilerServerLogger logger,
+            CancellationToken cancellationToken) =>
+            RunServerBuildRequestAsync(
+                buildRequest,
+                pipeName,
+                timeoutOverride: null,
+                tryCreateServerFunc: (pipeName, logger) => TryCreateServer(clientDirectory, pipeName, logger),
+                logger,
+                cancellationToken);
```

Comment:
```suggestion
        internal static Task<BuildResponse> RunServerBuildRequestAsync(
            BuildRequest buildRequest,
            string pipeName,
            string clientDirectory,
            ICompilerServerLogger logger,
            CancellationToken cancellationToken)
                => RunServerBuildRequestAsync(
                    buildRequest,
                    pipeName,
                    timeoutOverride: null,
                    tryCreateServerFunc: (pipeName, logger) => TryCreateServer(clientDirectory, pipeName, logger),
                    logger,
                    cancellationToken);
```
nit: having the method body at the same indent depth as the parameter list makes it a little more difficult to read. Also, in much of Roslyn, we tend to make the `=>` token for an expression body the first token on the line.